keps/sig-node/4381-dra-semantic-parameters/README.md

```diff
@@ -0,0 +1,1477 @@
+of each individual sub-partition.
+
+The "partitionable-gpu-1" GPU can be used either with the "full-gpu", "quarters",
+or "half-and-quarters" partitioning scheme. When using "half-and-quarters", there's
+another choice between a "half" and "quarters" partitioning scheme for the second half. Once a
+partitioning scheme has been chosen to satisfy a claim, it cannot be changed
+until all claims bound to resources from that partition have been deallocated.
+
+At the moment, resource attribute names and their type are defined by
+vendors. Resource names with ".k8s.io" as a suffix are reserved for future use
+and standardization by Kubernetes. This could be used to describe topology
+across resources from different vendors, but this is out-of-scope for now.
+
+If a driver needs to reduce resource capacity, then there is a risk that a
+claim gets allocated using that capacity while the kubelet is updating a
+NodeResourceSlice. The implementations of semantic models must handle scenarios
+where more resources are allocated than available. The kubelet plugin of a DRA
+driver ***must*** double-check that the allocated resources are still available
+when NodePrepareResource is called. If not, the pod cannot start until the
+resource comes back. Treating this as a fatal error during pod admission would
+allow us to delete the pod and trying again with a new one.
+
+### Using semantic parameters
+
+The following is an example CRD which the developer of the
+`cards.dra.example.com` DRA driver might define as a valid claim parameters
+object for requesting access to its GPUs:
+
+```yaml
+kind: CardParameters
+apiVersion: dra.example.com/v1alpha1
+metadata:
+  name: my-parameters
+  namespace: user-namespace
+  uid: foobar-uid
+...
+spec:
+  count: 2
+  minimumRuntimeVersion: v12.0.0
+  minimumMemory: 8Gi
+  # "sharing" is a configuration parameter that does not
+  # get translated into the selector below.
+  sharing:
+    strategy: TimeSliced
+```
+
+Note that all fields in this CRD can be fully validated since it is owned by
+the DRA driver itself. This includes value ranges that are specific to the
+underlying hardware. There's no risk of using invalid attribute names because
+only the fields shown here are valid.
+
+With this CRD in place, a DRA driver controller is able to convert instances of
+it into a generic, in-tree `ResourceClaimParameters` object that the scheduler
+is able to understand.
+
+For the example above, the converted object would look as follows:
+
+```yaml
+kind: ResourceClaimParameters
+apiVersion: resource.k8s.io/v1alpha2
+
+metadata:
+  # This cannot be the same as my-parameters because parameter objects with a different
+  # type might also use it. Instead, the original object gets linked to below.
+  name: someArbitraryName
+  namespace: user-namespace
+
+generatedFrom:
+  name: my-parameters
+  kind: CardParameters
+  apiGroup: dra.example.com
+  uid: foobar-uid
+
+vendorParameters:
```

Comment:
These are configuration parameters that are not needed for scheduling / CA decisions, but need to be passed to the kubelet to do configuration on the node as part of resource "preparation". I believe the idea was to use a `RawExtention` to represent it.

Reply:
Yes, exactly. So, the vendor CR contains various configuration parameters that are totally unrelated to scheduling and resource allocation. Those can remain opaque to K8s, but they need to make their way from the class and claim all the way down to the on-node driver. They also need to be captured *as they existed at the time of allocation*. So, the translation of the CR to vendor-neutral form also stuffs away any of these "config" related values from the class and claim into a passthrough RawExtension. This is then copied at the time of allocation into the resource claim status, along with the rest of the allocation information.

If a user subsequently updates the CR, it doesn't matter, because we have captured the config data as it was at the time of allocation. The updated CR will be used for future allocations but not for existing ones.

Reply:
hmm, I think I get it.  I don't love it, but I get it, I think.